fdb-extensions/src/main/java/com/apple/foundationdb/async/RankedSet.java

```diff
@@ -356,37 +396,59 @@ public RankedSet(Subspace subspace, Executor executor) {
                             if (count == null || count <= 0) {
                                 return READY_FALSE;
                             }
+                            // This works even if the current set does not track duplicates but duplicates were added
+                            // earlier by one that did.
+                            final boolean duplicate = count > 1;
                             final int nlevels = config.getNLevels();
                             final List<CompletableFuture<Void>> futures = new ArrayList<>(nlevels);
                             for (int li = 0; li < nlevels; ++li) {
                                 final int level = li;
-                                // This could be optimized to check the hash for which levels should have this key.
-                                final byte[] k = subspace.pack(Tuple.from(level, key));
 
                                 final CompletableFuture<Void> future;
-                                final CompletableFuture<byte[]> cf = tr.get(k);
 
-                                if (level == 0) {
-                                    future = cf.thenApply(c -> {
-                                        if (c != null) {
-                                            tr.clear(k);
-                                        }
+                                if (duplicate) {
+                                    // Always subtract one, never clearing a key. It is possible for this to leave
+                                    // a key with a count of zero, if duplicates were inserted with different hash functions.
+                                    Function<byte[], Void> decrement = k -> {
+                                        tr.mutate(MutationType.ADD, subspace.pack(Tuple.from(level, k)), encodeLong(-1));
```

Comment:
I see now that as we read the count and, if the count is 1, just remove the key, that this is no longer something to be worried about, though I suppose one could wist hopefully for a data structure with fewer conflicts (e.g., no conflicts when adding or removing the same key). It....might be possible, but it probably requires additional fanciness like conflicts in same "shadow realm", and it can be done later if we decide we really want it.